File: src/lib/commands.ts
/**
 * Typed wrappers around Tauri IPC invoke calls.
 */
import { invoke } from "@tauri-apps/api/core";

/** Port the panel's managed Astro dev server runs on (avoids conflicting with default 4321). */
export const DEV_SERVER_PORT = 4322;
export const DEV_SERVER_ORIGIN = `http://localhost:${DEV_SERVER_PORT}`;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AppConfig {
  repo_path: string | null;
  theme: string | null;
  cf_account_id: string | null;
  cf_project_name: string | null;
  cf_api_token: string | null;
  cf_domain: string | null;
  cf_zone_id: string | null;
}

export interface ContentEntry {
  slug: string;
  content_type: "post" | "app";
  title: string;
  summary: string;
  tags: string[];
  is_draft: boolean;
  created_date: string;
  publication_date: string | null;
  author: string | null;
  image: string | null;
  file_path: string;
  modified_date: string | null;
  published_hash: string | null;
  has_changed: boolean;
}

export interface HealthStatus {
  url: string;
  ok: boolean;
  status_code: number | null;
}

export interface CfDeploymentInfo {
  deployed_at: string;
  commit_hash: string | null;
  commit_message: string | null;
  status: string;
  url: string | null;
}

export interface CfDailyCount {
  date: string;
  count: number;
  uniques: number;
  bytes: number;
  cached_bytes: number;
  cached_requests: number;
  threats: number;
}

export interface CfPathCount {
  path: string;
  count: number;
}

export interface CfCountryCount {
  country: string;
  count: number;
}

export interface CfStatusCount {
  status: number;
  count: number;
}

export interface CfBrowserCount {
  browser: string;
  page_views: number;
}

export interface CfAnalytics {
  period: string;
  total_requests: number;
  daily_requests: CfDailyCount[];
  top_paths: CfPathCount[];
  top_countries: CfCountryCount[];
  status_codes: CfStatusCount[];
  browsers: CfBrowserCount[];
}

export interface CreatePostArgs {
  title: string;
  slug: string;
  summary: string;
  tags: string[];
}

export interface CreateAppArgs {
  title: string;
  slug: string;
  summary: string;
  tags: string[];
}

// ---------------------------------------------------------------------------
// Commands
// ---------------------------------------------------------------------------

export function getConfig(): Promise<AppConfig> {
  return invoke("get_config");
}

export function setConfig(config: AppConfig): Promise<void> {
  return invoke("set_config", { config });
}

export function validateRepoPath(path: string): Promise<boolean> {
  return invoke("validate_repo_path", { path });
}

export function listContent(repoPath: string): Promise<ContentEntry[]> {
  return invoke("list_content", { repoPath });
}

export function readFile(path: string): Promise<string> {
  return invoke("read_file", { path });
}

export function writeFile(path: string, content: string): Promise<void> {
  return invoke("write_file", { path, content });
}

export function createPost(repoPath: string, args: CreatePostArgs): Promise<ContentEntry> {
  return invoke("create_post", { repoPath, args });
}

export function createApp(repoPath: string, args: CreateAppArgs): Promise<ContentEntry> {
  return invoke("create_app", { repoPath, args });
}

export function deleteContent(repoPath: string, slug: string): Promise<void> {
  return invoke("delete_content", { repoPath, slug });
}

export function publish(repoPath: string, slug: string): Promise<ContentEntry> {
  return invoke("publish", { repoPath, slug });
}

export function unpublish(repoPath: string, slug: string): Promise<ContentEntry> {
  return invoke("unpublish", { repoPath, slug });
}

export function rollback(repoPath: string, slug: string): Promise<ContentEntry> {
  return invoke("rollback", { repoPath, slug });
}

export function gitStatus(repoPath: string): Promise<string> {
  return invoke("git_status", { repoPath });
}

export function openInVscode(path: string): Promise<void> {
  return invoke("open_in_vscode", { path });
}

export function startWatcher(repoPath: string): Promise<void> {
  return invoke("start_watcher", { repoPath });
}

export function stopWatcher(): Promise<void> {
  return invoke("stop_watcher");
}

export function startDevServer(repoPath: string): Promise<void> {
  return invoke("start_dev_server", { repoPath });
}

export function stopDevServer(): Promise<void> {
  return invoke("stop_dev_server");
}

export function checkUrlHealth(url: string): Promise<HealthStatus> {
  return invoke("check_url_health", { url });
}

export function fetchLastDeployment(): Promise<CfDeploymentInfo> {
  return invoke("fetch_last_deployment");
}

export function fetchAnalytics(days: number, engagement: boolean): Promise<CfAnalytics> {
  return invoke("fetch_analytics", { days, engagement });
}

export function testCfConnection(): Promise<string> {
  return invoke("test_cf_connection");
}

File: src-tauri/src/commands.rs
use std::fs;
use std::path::Path;
use std::process::Command;
use std::time::Duration;
use tauri::Manager;

use crate::config;
use crate::content;
use crate::frontmatter;
use crate::git;
use crate::security;
use crate::types::*;

// ---------------------------------------------------------------------------
// Config commands
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn get_config(app: tauri::AppHandle) -> Result<AppConfig, String> {
    Ok(config::load_config(&app))
}

#[tauri::command]
pub fn set_config(app: tauri::AppHandle, config: AppConfig) -> Result<(), String> {
    config::save_config(&app, &config)
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn validate_repo_path(path: String) -> Result<bool, String> {
    let base = Path::new(&path);
    let has_blog = base.join("src/content/blog").is_dir();
    let has_apps = base.join("src/content/apps").is_dir();
    let has_pkg = base.join("package.json").is_file();
    Ok(has_blog && has_apps && has_pkg)
}

// ---------------------------------------------------------------------------
// Content commands
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn list_content(repo_path: String) -> Result<Vec<ContentEntry>, String> {
    content::list_content(&repo_path)
}

#[tauri::command]
pub fn read_file(app: tauri::AppHandle, path: String) -> Result<String, String> {
    let cfg = config::load_config(&app);
    if let Some(ref repo_path) = cfg.repo_path {
        let base = Path::new(repo_path);
        let target = Path::new(&path);
        security::ensure_within(target, base)?;
    }
    fs::read_to_string(&path).map_err(|e| format!("Failed to read {}: {}", path, e))
}

#[tauri::command]
pub fn write_file(app: tauri::AppHandle, path: String, content: String) -> Result<(), String> {
    let cfg = config::load_config(&app);
    if let Some(ref repo_path) = cfg.repo_path {
        let base = Path::new(repo_path);
        let target = Path::new(&path);
        security::ensure_within(target, base)?;
    }
    fs::write(&path, content).map_err(|e| format!("Failed to write {}: {}", path, e))
}

#[tauri::command]
pub fn create_post(repo_path: String, args: CreatePostArgs) -> Result<ContentEntry, String> {
    content::create_post(&repo_path, args)
}

#[tauri::command]
pub fn create_app(repo_path: String, args: CreateAppArgs) -> Result<ContentEntry, String> {
    content::create_app(&repo_path, args)
}

#[tauri::command]
pub async fn delete_content(repo_path: String, slug: String) -> Result<(), String> {
    content::delete_content(&repo_path, &slug)
}

// ---------------------------------------------------------------------------
// Publish / unpublish
// ---------------------------------------------------------------------------

#[tauri::command]
pub async fn publish(repo_path: String, slug: String) -> Result<ContentEntry, String> {
    let base = Path::new(&repo_path);
    let (file_path, content_type) = content::find_content_file(base, &slug)?;
    let file_content = fs::read_to_string(&file_path).map_err(|e| e.to_string())?;
    let (yaml, rest) = frontmatter::split_frontmatter(&file_content)
        .ok_or_else(|| "Could not parse frontmatter.".to_string())?;

    if frontmatter::get_yaml_bool(&yaml, "isDraft") == Some(false) {
        return frontmatter::parse_content_entry(&slug, &content_type, &file_path)
            .ok_or_else(|| "Failed to parse entry.".to_string());
    }

    let mut new_yaml = frontmatter::set_frontmatter_field(&yaml, "isDraft", "false");

    let has_pub_date = frontmatter::PUB_DATE_RE.is_match(&new_yaml);
    if !has_pub_date {
        new_yaml = frontmatter::insert_field_after(
            &new_yaml,
            "createdDate",
            "publicationDate",
            &format!("\"{}\"", frontmatter::now_iso()),
        );
    }

    let new_content = frontmatter::assemble_file(&new_yaml, &rest);
    fs::write(&file_path, &new_content).map_err(|e| e.to_string())?;

    let rel_path = file_path
        .strip_prefix(base)
        .unwrap_or(&file_path)
        .to_string_lossy()
        .to_string();
    let title = frontmatter::get_yaml_field(&new_yaml, "title").unwrap_or_else(|| slug.clone());
    git::git_add_commit_push(&repo_path, &rel_path, &format!("publish: {}", title))?;

    frontmatter::parse_content_entry(&slug, &content_type, &file_path)
        .ok_or_else(|| "Failed to parse entry after publish.".to_string())
}

#[tauri::command]
pub async fn unpublish(repo_path: String, slug: String) -> Result<ContentEntry, String> {
    let base = Path::new(&repo_path);
    let (file_path, content_type) = content::find_content_file(base, &slug)?;
    let file_content = fs::read_to_string(&file_path).map_err(|e| e.to_string())?;
    let (yaml, rest) = frontmatter::split_frontmatter(&file_content)
        .ok_or_else(|| "Could not parse frontmatter.".to_string())?;

    if frontmatter::get_yaml_bool(&yaml, "isDraft") == Some(true) {
        return frontmatter::parse_content_entry(&slug, &content_type, &file_path)
            .ok_or_else(|| "Failed to parse entry.".to_string());
    }

    let new_yaml = frontmatter::set_frontmatter_field(&yaml, "isDraft", "true");
    let new_content = frontmatter::assemble_file(&new_yaml, &rest);
    fs::write(&file_path, &new_content).map_err(|e| e.to_string())?;

    let rel_path = file_path
        .strip_prefix(base)
        .unwrap_or(&file_path)
        .to_string_lossy()
        .to_string();
    let title = frontmatter::get_yaml_field(&new_yaml, "title").unwrap_or_else(|| slug.clone());
    git::git_add_commit_push(&repo_path, &rel_path, &format!("unpublish: {}", title))?;

    frontmatter::parse_content_entry(&slug, &content_type, &file_path)
        .ok_or_else(|| "Failed to parse entry after unpublish.".to_string())
}

// ---------------------------------------------------------------------------
// Git
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn git_status(repo_path: String) -> Result<String, String> {
    git::git_status_porcelain(&repo_path)
}

// ---------------------------------------------------------------------------
// VS Code
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn open_in_vscode(path: String) -> Result<(), String> {
    if Command::new("code").arg(&path).spawn().is_ok() {
        return Ok(());
    }

    #[cfg(target_os = "macos")]
    {
        Command::new("open")
            .args(["-a", "Visual Studio Code", &path])
            .spawn()
            .map_err(|e| format!("Failed to open VS Code: {}", e))?;
        Ok(())
    }

    #[cfg(not(target_os = "macos"))]
    Err("VS Code CLI ('code') not found in PATH.".to_string())
}

// ---------------------------------------------------------------------------
// URL health check
// ---------------------------------------------------------------------------

#[tauri::command]
pub async fn check_url_health(url: String) -> Result<HealthStatus, String> {
    if !url.starts_with("http://") && !url.starts_with("https://") {
        return Err("Only HTTP(S) URLs are allowed".to_string());
    }
    let lower = url.to_lowercase();
    let is_dev_server = lower.starts_with("http://localhost:4322");
    if !is_dev_server
        && (lower.contains("://localhost")
            || lower.contains("://127.")
            || lower.contains("://0.0.0.0")
            || lower.contains("://[::1]")
            || lower.contains("://10.")
            || lower.contains("://192.168."))
    {
        return Err("Requests to local/private addresses are not allowed".to_string());
    }

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(5))
        .build()
        .map_err(|e| e.to_string())?;

    match client.head(&url).send().await {
        Ok(resp) => {
            let status = resp.status().as_u16();
            Ok(HealthStatus {
                url,
                ok: resp.status().is_success(),
                status_code: Some(status),
            })
        }
        Err(_) => Ok(HealthStatus {
            url,
            ok: false,
            status_code: None,
        }),
    }
}

// ---------------------------------------------------------------------------
// Watcher & dev server (delegates to existing modules)
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn start_watcher(app: tauri::AppHandle, repo_path: String) -> Result<(), String> {
    crate::watcher::start_watching(&app, &repo_path)
}

#[tauri::command]
pub fn stop_watcher(app: tauri::AppHandle) -> Result<(), String> {
    crate::watcher::stop_watching(&app)
}

#[tauri::command]
pub fn start_dev_server(app: tauri::AppHandle, repo_path: String) -> Result<(), String> {
    let state = app.state::<crate::devserver::DevServerState>();
    crate::devserver::start_dev_server(&state, &repo_path)
}

#[tauri::command]
pub fn stop_dev_server(app: tauri::AppHandle) -> Result<(), String> {
    let state = app.state::<crate::devserver::DevServerState>();
    crate::devserver::stop_dev_server(&state)
}
